### ISSUE TITLE
[RFC] GraphQL Input Union type

### ISSUE THREAD

**Issue by frikille** [graphql/graphql-spec#488]
[RFC] GraphQL Input Union type
------

## Background

There have been many threads on different issues and pull requests regarding this feature. The main discussion started with the graphql/graphql-js#207 on 17/10/2015.

Currently, GraphQL input types can only have one single definition, and the more adoption GraphQL gained in the past years it has become clear that this is a serious limitation in some use cases. Although there have been multiple proposals in the past, this is still an unresolved issue.

With this RFC document, I would like to collect and summarise the previous discussions at a common place and proposals and propose a new alternative solution.

To have a better understanding of the required changes there is a [reference implementation](https://github.com/mozaikio/graphql-js/tree/input-union-rfc-implementation) of this proposal, but that I will keep up to date based on future feeback on this proposal.

The following list shows what proposals have been put forward so far including a short summary of pros and cons added by @treybrisbane in [this comment](https://github.com/facebook/graphql/pull/395#issuecomment-361784141)

* **`__inputname` field** by @tgriesser

  RFC document: #395 

  Reference implementation: graphql/graphql-js#1196

  This proposal was the first official RFC which has been discussed at the last [GraphQL Working Group meeting](https://github.com/graphql/graphql-wg/blob/master/notes/2018-06-14.md#discuss-proposed-rfc-for-inputunion-type-related-graphql-js-pr-tg-10m).
  This proposal in this current form has been rejected by the WG because of the `__inputname` semantics. However, everyone agrees that alternative proposals should be explored.

  * Pros:
    * Expresses the design intent within the schema
    * Supports unions of types with overlapping fields
    * Removes the costs of the tagged union pattern (both to the schema and field resolution)
    * Addresses the feature asymmetry of unions within the type system
  * Cons:
    * Adds complexity to the language in the form of input union-specific syntax
    * Adds complexity to the language in the form of additional validation (around __inputtype, etc)
    * Adds complexity to the request protocol in the form of a (input union-specific) constraint

* **Tagged union** by @leebyron and @IvanGoncharov 
  
  [Original comment](https://github.com/facebook/graphql/pull/395#issuecomment-361373097)

  ```graphql
  input MediaBlock = { post: PostInput! } | { image: ImageInput! }
  ```

  * Pros:
    * Expresses the design intent within the schema
    * Removes the costs of the tagged union pattern (both to the schema and field resolution)
    * Does not require changes to the request protocol
  * Cons:
    * No support for unions of types with overlapping fields
    * Introduces inconsistencies between the syntax for output unions and the syntax for input unions
    * Adds complexity to the language in the form of input union-specific syntax
    * Adds complexity to the language in the form of additional validation (around enforcing the stipulations on overlapping fields, nullability, etc)
    * Does not fully address the feature asymmetry of unions within the type system

* **Directive**
  
  [Original comment](https://github.com/facebook/graphql/pull/395#issuecomment-361674099)

  ```graphql
  input UpdateAccountInput @inputUnion {
    UpdateAccountBasicInput: UpdateAccountBasicInput
    UpdateAccountContactInput: UpdateAccountContactInput
    UpdateAccountFromAdminInput: UpdateAccountFromAdminInput
    UpdateAccountPreferencesInput: UpdateAccountPreferencesInput
  }
  ```

  * Pros:
    * Requires no language or request prototcol changes beyond a new directive
    * Supports unions of types with overlapping fields
  * Cons:
    * Adds complexity to the language in the form of a new directive
    * Does not express the design intent within the schema (the presence of a directive completely changes the meaning of a type definition which would otherwise be used to describe a simple object type)
    * Does not remove the costs of the tagged union pattern
    * Does not address the feature asymmetry of unions within the type system

## Proposed solution

Based on the previous discussions I would like to propose an alternative solution by using a disjoint (or discriminated) union type.

Defining a disjoint union has two requirements:
1. Have a common literal type field - called the _discriminant_
2. A type alias that takes the union of those types - the _input union_

With that our GraphQL schema definition would be the following ([Full schema definition](https://gist.github.com/frikille/8220a77049b617db553309f001ca8a85))

```graphql
literal ImageInputKind
literal PostInputKind

input AddPostInput {
  kind: PostInputKind!
  title: String!
  body: String!
}

input AddImageInput  {
  kind: ImageInputKind!
  photo: String!
  caption: String
}

inputUnion AddMediaBlock = AddPostInput | AddImageInput

input EditPostInput {
  inputTypeName: PostInputKind!
  title: String
  body: String
}

input EditImageInput {
  inputTypeName: ImageInputKind!
  photo: String
  caption: String
}

inputUnion EditMediaBlock = EditPostInput | EditImageInput

type Mutation {
  createMedia(content: [AddMediaBlock]!): Media   
  editMedia(content: [EditMediaBlock]!): Media
}
```

And a mutation query would be the following:

```graphql
mutation {
  createMedia(content: [{
    kind: PostInputKind
    title: "My Post"
    body: "Lorem ipsum ..."
  }, {
    kind: ImageInputKind
    photo: "http://www.tanews.org.tw/sites/default/files/imce/admin/user2027/6491213611_c4fc290a33_z.jpg"
    caption: "Dog"
  }]) {
    content {
      ... on PostBlock {
        title
        body
      }
      ... on ImageBlock {
        photo
        caption
      }
    }
  }
}
```

or with variables
```graphql
mutation CreateMediaMutation($content: [AddMediaBlock]!) {
  createMedia(content: $content) {
    id
    content {
      ... on PostBlock {
        title
        body
      }
      ... on ImageBlock {
        photo
        caption
      }
    }
  }
}
```

```js
{
  "content": [{
    "kind": "PostInputKind",
    "title": "My Post",
    "body": "Lorem ipsum ...",
  }, {
    "kind": "ImageInputKind",
    "photo": "http://www.tanews.org.tw/sites/default/files/imce/admin/user2027/6491213611_c4fc290a33_z.jpg",
    "caption": "Dog"
  }]
}
```

#### Literal type
The GraphQLLiteralType is an exact value type. This new type enables the definition a discriminant field for input types that are part of an input union type.

#### Input union type
The GraphQLInputUnionType represent an object that could be one of a list of GraphQL Input Object types.

**Input Coercion**

The input union type needs a way of determining which input object a given input value correspond to. Based on the discriminant field it is possible to have an internal function that can resolve the correct GraphQL Input Object type. Once it has been done, the input coercion of the input union is the same as the input coercion of the input object.

**Type Validation**

Input union types have a potential to be invalid if incorrectly defined:

- An input union type must include one or more unique input objects
- Every included input object type must have:
  - One common field
  - This common field must be a unique Literal type for every Input Object type

### Using String or Enum types instead of Literal types

While I think it would be better to add support for a Literal type, I can see that this type would only be useful for Input unions; therefore, it might be unnecessary.  However, it would be possible to use String or Enum types for the discriminant field, but in this case, a `resolveType` function must be implemented by the user. This would also remove one of the type validations required by the input type (2.2 - The common field must be a unique Literal type for every Input Object type).

## Final thoughts

I believe that the above proposal addresses the different issues that have been raised against earlier proposals. It introduces additional syntax and complexity but uses a concept (disjoint union) that is widely used in programming languages and type systems. And as a consequence  I think that the pros of this proposal outweigh the cons.

* **Pros**
  * Expresses the design intent within the schema
  * The discriminant field name is configurable
  * Supports unions of types with overlapping fields
  * Addresses the feature asymmetry of unions within the type system
  * Does not require changes to the request protocol

* **Cons**
  * Adds complexity to the language in the form of literal-specific syntax (might not need)
  * Adds complexity to the language in the form of input union-specific syntax
  * Adds complexity to the language in the form of additional validations (input union, discriminant field, input union resolving (might not need))

However, I think there are many questions that needs some more discussions, until a final proposal can be agreed on - especially around the new literal type, and if it is needed at all.

**Comment by IvanGoncharov** [comment]
> Currently, GraphQL input types can only have one single definition, and the more adoption GraphQL gained in the past years it has become clear that this is a serious limitation in some use cases. 

This is the main point of this discussion, instead of proposing competing solutions we should focus on defining a set of problems that we want to address. From previous discussions, it looks like the main use case is "super" mutations. If this is also the case for you RFC should address why you can't do separate `createMediaPost` or `createMediaImage`. Do you have any other use real-life use cases that you want to address with this feature?

So this discussion should be driven by real-life user cases and the proposed solution should be presented only in the context of solving this use cases.
IMHO, without a defined set of the use cases, this discussion looks like a beauty contest for SDL syntax and JSON input.

> Does not address the feature asymmetry of unions within the type system

Feature asymmetry between input and output types is not an issue since they are fundamentally different in API case. 

Output types represent entities from the business domain (e.g. `Post`, `Image`, etc.) and because of that, they are usually reused all around the schema.

Input types represent set of data required for a particular action or query (e.g. `AddPostInput`, `EditPostInput`) and thus in the majority of cases, input types have very limited reusability.

IMHO, that's why it makes sense to sacrifice some simplicity to allow better expressing your domain entities and facilitate reusability by allowing interfaces and unions on output types.

As for input types instead of the feature-reach type system, we need to provide set of validation constraints, e.g. based on your example this type will satisfy:
```graphql
type AddMediaBlock {
  kind: PostInputKind!
  title: String
  body: String
  photo: String
  caption: String
}
```

It's working and fully solve the present problem there are the only two problems:
* documentation - you need to describe valid field combination in `description`
* validation - you need to write validation manually

So instead of creating two additional types and complicating type system with `inputUnion` we can just solve the above problems by describing required validation constraints: `type AddMediaBlock = { ... } | { ... }`.

> Using String or Enum types instead of Literal types

IMHO, `literal` is just polute namespace and force to use `PostInputKind` istead of `Post`. 
We shouldn't mix client data with typenames from typesystem in any shape or form especially since it was one of main arguments against `__inputname`.

So it should be either in place defined strings or enum values.

**Comment by frikille** [comment]
> From previous discussions, it looks like the main use case is "super" mutations.
> Do you have any other use real-life use cases that you want to address with this feature?

Yes. We are building a headless content management platform: users can define their own content types, fields and valiations. They get a GraphQL API that they can use to read and write data. And we want to support a new feature whereby a content type field can have multiple types. So an output type would be something like this:

```graphql
type PageDocument {
  id: ID
  title: String
  slug: String
  author: Author
  sections: [Section]
}

type Section {
  title: String
  blocks: [SectionBlock]
}

union SectionBlock = ImageGallery | HighlightedImage | ArticleList | RichText
```

As you can see, this doesn't requires a top level "super" mutation. I agree with you on that a super mutation can and should be just separate mutations. But in our case it is not not possible since the input field that can have multiple types is deeply nested.

An example schema for the input types:

```graphql
mutation createPageDocument(document: PageDocumentInput!): PageDocument

input PageDocumentInput {
  title: String!
  slug: String!
  author: ID!
  sections: [SectionInput]
}

input SectionInput = {
  title: String!
  blocks: [SectionBlockInput]
}

enum SectionBlockInputEnum = {
  ImageGallery
  HighlightedImage
  Articles
  RichText
}

input ImageGalleryInput = {
  kind: SectionBlockInputTypeEnum!
  title: String!
  description: String!
  imageUrls: [String!]!
}

input HighlightedImageInput {
  kind: SectionBlockInputTypeEnum!
  url: String!
  caption: String!
}

input ArticaleListInput {
  kind: SectionBlockInputTypeEnum!
  articleIds: [ID!]!
}

input RichTextInput {
  kind: SectionBlockInputTypeEnum!
  html: String
  markdown: String
}
```

I deliberately left out the definition of the `SectionBlockInput` input type. Based on my proposal it would be this:
```graphql
inputUnion SectionBlockInput = ImageGalleryInput | HighlightedImageInput | ArticleListInput | RichTextInput
```

I would skip the "super" `SectionBlockInput` type definition, that includes every field from the four input types, because as you mentioned that have two problems:
* documentation: We must be able to generate the best documentation to our users since they can use the API for writing data in to their project
* validation: this is the bigger issue. We started with GraphQL mostly because of the type system and query validation it provides. And while we can write a validation engine that solves this problem, we feel that it's just another validation should live in GraphQL.

With your required validation constraint suggestion I suppose it would be something like this:
```graphql
input SectionBlockInput = {  
  kind: SectionBlockInputTypeEnum!
  title: String!
  description: String!
  imageUrls: [String!]!
} | {
  kind: SectionBlockInputTypeEnum!
  url: String!
  caption: String!
} | {
  kind: SectionBlockInputTypeEnum!
  articleIds: [ID!]!
} | {
  kind: SectionBlockInputTypeEnum!
  html: String
  markdown: String
}
```

I think these two examples are pretty close to each other and would achieve the same result. The main question is that it would be better doing this:
* with a new type (inputType)
* or by new validation constraint

> IMHO, literal is just polute namespace and force to use `PostInputKind` istead of `Post`.
> We shouldn't mix client data with typenames from typesystem in any shape or form especially since it was one of main arguments against `__inputname`.

I accept this.It had a lot of advantages using it when I implemented the input type resolver. I would be happy to drop this one completely as it already felt a bit unnecessary.

**Comment by IvanGoncharov** [comment]
@frikille Thank you for sharing your use case it helped me a lot with understanding your particular problem. 
Is it the only type of scenarios where you need Input Union types?

**Comment by frikille** [comment]
There are some other places where it would be useful, but those are top level mutations (creating a document) which can be solved other ways. However the input union support would help with that use case as well. Our main problem is that we have completely different schemas for every user project, and it would be better if we were able to use a common mutation for creating a document instead of dynamically changing the mutation name. But I understand that it can be solved without input unions, and we've been using that solution for a long time so for that it is not necessary. However, we don't see how the use case I described in my previous comment can be solved without input unions.

**Comment by jturkel** [comment]
Creating/updating an entity (`PageDocument` in the example above) with a heterogeneous collection of nested value objects (different types of `Section` in the example above) is a great use case for input unions. If you're looking for more use cases @xuorig describes batching multiple mutating operations (each of which may have very different structure) into a single atomic mutation in [this](https://medium.com/@__xuorig__/graphql-mutation-design-batch-updates-ca2452f92833) blog post. The post describes a project board like Trello where a user can add/move/delete/update cards in a single atomic operation. The corresponding SDL looks something like this with input unions:

```gql
type Mutation {
  updateBoard(input: UpdateBoardInput!): UpdateBoardPayload!
}

type UpdateBoardInput {
  operations: [Operation!]!
}

union Operation = AddOperation | UpdateOperation |  MoveOperation | DeleteOperation

input AddOperation {
  id: ID!
  body: String!
}

input UpdateOperation {
  id: ID!
  body: String!
}

input MoveOperation {
  id: ID!
  newPos: Int!
}

input DeleteOperation {
  id: ID!
  body: String!
}
```
@OlegIlyenko also described similar ideas for batching multiple operations into a single atomic mutation in the context of CQRS and commerce in [this](https://techblog.commercetools.com/modeling-graphql-mutations-52d4369f73b1) blog post.

**Comment by jturkel** [comment]
@IvanGoncharov - We have examples in the product information management space where input unions would be useful as field arguments. On such example is retrieving a filtered list of products: 

```gql
input StringEqualityFilter {
  propertyId: String!
  value: String!
}

input NumericRangeFilter {
  propertyId: String!
  minValue: Int!
  maxValue: Int!
}

input HasNoValueFilter {
  propertyId: String!
}

inputunion Filter = StringEqualityFilter | NumericRangeFilter | HasNoValueFilter

type Query {
  products(filters: [Filter!]): ProductCursor
}
```

Given this schema customers can retrieve a list of products that meet the following conditions:

* The 'Brand' property is equal to 'Apple'
* The 'Category' property is equal 'Laptops'
* The 'Price' property is between 1000 and 2000
* The 'Description' property has no value

Note our customers can create their own "schema" for modelling product information so our GraphQL schema can't have hardcoded knowledge of things like brands and categories thus the more general `propertyId` in this example.

**Comment by treybrisbane** [comment]
@frikille Thanks for writing this up. :)

I definitely see why you went with the literal type. What was the rational behind going with a separate `literal Foo` declaration rather than just this:
```graphql
input Foo {
  kind: "Foo"!
  # ...etc
}
```
???

**Comment by frikille** [comment]
@treybrisbane We were thinking about that and we think that supporting inline strings would be a nice syntax sugar but at the end, it would be parsed to a new type, so because of clarity we added the literal type (also I kinda prefer to declare everything properly)

@IvanGoncharov I'm not that familiar with Java any more, but it is a good point, and as far as I know, it doesn't have support for discriminated unions.

**Comment by IvanGoncharov** [comment]
> I'm not that familiar with Java any more, but it is a good point, and as far as I know, it doesn't have support for discriminated unions.

Mobile is a pretty critical platform for GraphQL, so it would be great to know how it would be affected. So would be great if someone knowledgeable can answer my previous question:
> how proposed input unions affect code generation especially for strong type languages, e.g. Java. I suspect that such languages don't support unions based on field value as the discriminator. It especially important for mobile clients.
Can someone check how it will affect apollo-ios and apollo-android?

**Comment by sorenbs** [comment]
I am very happy that progress is being made on union input types 🎉 

However, if I understand the proposal correctly, then it will not address the use cases we have in mind for the Prisma API.

Here is a spec for a new feature we will be implementing soon. I have written up how we would like to shape the API if union input types would support it. I have also described why the proposed solution with a discriminating field does not work for us: https://github.com/prismagraphql/prisma/issues/1349

I'm curious to hear if others are seeing use cases similar to us, or if we are special for implementing a very generic API?

Let me know if more details are needed.

**Comment by dendrochronology** [comment]
@sorenbs would you mind explaining your reasoning a bit more, or perhaps provide some basic query examples, please? Perhaps I'm misunderstanding your atomic actions proposal, but it seems much broader than what's being discussed here. 

I think the core issue this RFC addresses is to provide a simple, declarative way to allow for multiple input types on query and mutation fields. That's a severe weakness of GraphQL when it comes to data modeling.

A lot of my recent work has been porting data from legacy relational CMS's into modern GraphQL environments, and this RFC would solve one of the biggest headaches I've had. I'd rather see the community focus on feature-lean solutions to the biggest gaps first (i.e., things without good workarounds), and focus on narrower issues in later iterations.

**Comment by sorenbs** [comment]
@dendrochronology - sorry for just linking to a giant feature spec, thats certainly not the best way to illustrate my point :-)

We have evolved our generic GraphQL API over the last 3 years. At 4-5 times throughout this process has it been brought up that some form of union input type could simplify the API design. Most of these cases boils down to supporting either a specific scalar type or a specific complex type:

given the types

```graphql
type Complex {
  begin: DateTime
  end: DateTime
}

union scalarOrComplex = Int | Complex

type mutation {
  createTiming(duration: scalarOrComplex): Int
}
```

Create a timing providing duration in milliseconds or as a complex structure containing start and end DateTime:

```graphql
mutation {
  simple: createTiming(duration: 120)
  complex: createTiming(duration: {begin: "2017", end: "2018"})
}
```

For us it is paramount that we maintain a simple API.

Did this example make it clear what we are looking to achieve?

It might also be worth noting that we have found it valuable on many occasions that A single element is treated the same as an array with a single element. This flexibility has made it easier to evolve our API in many cases, and I think we can achieve something similar for union input types if we choose a system that solves for that.

From this perspective, we would prefer the `Tagged union by @leebyron and @IvanGoncharov` described above. Keep in mind that I have not thought through the implications for tooling and strongly typed languages.

**Comment by dendrochronology** [comment]
Thanks, @sorenbs, those are great examples. I do like the simplicity, and agree that robust APIs should handle singletons vs. arrays elegantly, a 'la [Postel's Law](https://en.wikipedia.org/wiki/Robustness_principle).

Back to @IvanGoncharov's request for a use case-driven discussion, I think that the union type implementation needs to go a bit further than generic APIs. A lot of my work is content and UX-based, where a client application pushes/pulls lots of heterogeneous objects from a back end, and displays them to a user for viewing/interaction.

For example, think of a basic drag-and-drop CMS interface, where a user can compose multiple blocks, with some levels of nesting and validation rules (e.g., block type X can have children of types A and B, but not C, etc.). With the current state of GraphQL, you need an awful lot of client-side logic to pre/post process output, because schemas have many fields with no baked-in validation or guarantees of meeting a required contract for available data or metadata fields (e.g., all blocks should have at least a type, label, color, x and y coordinates, ordinal number, etc.). I can mockup something like that in React easily, but building a GraphQL API to power it quickly gets messy.

I suppose my key takeaway is: I find that the current GraphQL spec makes it difficult to model a lot of real-world data situations without excessive verbosity and generous additional processing. I would _love_ to be wrong, and have someone point out what I'm missing, but that hasn't happened yet 🤔

**Comment by jturkel** [comment]
@IvanGoncharov - The proposed input union functionality for strongly typed clients should behave much the same way as "output" 
unions. For Java, which doesn't support unions, this can be modeled via an interface representing the union that each member type implements. Simplified code for earlier `createMedia` mutation would look something like:

```java
public interface AddMediaBlock { }

public class AddPostInput implements AddMediaBlock { }

public class AddImageInput implements AddMediaBlock { }

public class CreateMediaMutation {
  public CreateMediaMutation(AddMediaBlock[] content) { }
}

// Sample code that instantiates the appropriate union member types
AddMediaBlock[] content = new AddMediaBlock[]{new AddPostInput(...), new AddPostInput(...), new AddImageInput(...)};
client.mutate(new CreateMediaMutation(content));
```
Presumably Swift uses a similar strategy with [protocols](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html) (the moral equivalent of Java interfaces).

**Comment by jturkel** [comment]
@IvanGoncharov - What's your opinion on how to effectively move input unions forward? Seems like this issue has a pretty good catalog of use cases now and it's a matter of hammering out the technical details of input unions. It would be great to get a champion from the GraphQL WG to help guide this through the spec process. Is this something you have the time and desire to take on since you've already been pretty involved? I'm happy to help out in anyway I can (and I suspect others are too) but I'd like to make sure anything I do is actually helpful :)

**Comment by jturkel** [comment]
Great to hear you're up for helping out with this @IvanGoncharov! Figuring out where folks agree/disagree sounds like a perfect next step to me. Feel free to ping me on the GraphQL Slack if there's anything I can do to help.

**Comment by IvanGoncharov** [comment]
@jturkel Update: Have an urgent work so I wouldn't be able to compile something until next weekends.

**Comment by dmitrif** [comment]
Any updates on this? We are facing a similar need, unless there's a better way one may suggest:

```input LocationQueryIdOptions {
  id: Int!
}

input LocationQueryCoordsOptions {
  latitude: Int!
  longitude: Int!
}

union LocationOptions = LocationQueryIdOptions | LocationQueryCoordsOptions;

type Query {
  location(options: LocationQueryIdOptions | LocationQueryCoordsOptions): Location
}
```

**Comment by amccloud** [comment]
@dmitrif 👍 that's pretty much exactly what I tried before finding this issue:


```graphql
input PostDescriptor {
  # ...
}

input ImageDescriptor {
  # ...
}

union ObjectDescriptor = PostDescriptor | ImageDescriptor;

type Mutation {
  Comment(objectDescriptor: ObjectDescriptor!, comment: CommentInput): CommentMutations
}
```

**Comment by solussd** [comment]
+1 from me. Specifically, being able to atomically add/update entities via a single mutation that have heterogeneous collections (e.g., conforming to an interface or union definition).

**Comment by bericp1** [comment]
Another use case for this that we've encountered is recursive tree structures where subtrees can either be another tree of the same type as the root or a leaf type. As in:

```gql
input Tree {
  children: [Subtree!]!
  someRequiredTreeStringProperty: String!
  # ...
}

input Leaf {
  someRequiredLeafStringProperty: String!
  # ...
}

inputUnion Subtree = Tree | Leaf

type Query {
  someQuery(inputTree: Tree): SomeType
}
```

Currently – unless I'm missing something and short of defining a custom scalar type for our specific implementation – we have to less-than-ideally define the tree and leaf as a single type and differing fields that should technically be Non-Nullable must be left Nullable and then validated in the resolver. Detection and validation of whether or not a value in `children` is a `Tree` or a `Leaf` also is left up to the resolver (either by looking at the present fields or using some sort of `type` field):

```gql
input TreeOrLeaf {
  children: [TreeOrLeaf!]
  someRequiredTreeStringProperty: String
  someRequiredLeafStringProperty: String
}

type Query {
  someQuery(inputTree: TreeOrLeaf): SomeType
}
```

**Comment by alfaproject** [comment]
Just stumbled upon this RFC looking for a way to have a union input for an authentication API whereby I have multiple providers.

I guess for now I'll just create a generic authenticate input accepting all arguments for all providers and then validate the input fields on the back-end but it feels lame. ):

**Comment by jonaskello** [comment]
Our use case is "batching multiple operations into a single atomic mutation" from above. The user can make several changes then save, and we need to handle that as an atomic SQL transaction. I'm not sure how to handle a SQL transaction in the mutation resolver if we cannot get all the operations as an array input for a single mutation field. The only work-around I can think of is an array of a mega-type with all possible operations as fields, and only set one field per mega-type instance.

**Comment by raderio** [comment]
Our use cases are:
- registration: user can register by email/password or by social network
- create a product: it can be a _draft_ or _published_, also as second level it has type that can _good_ or _service_.

So, will be grate to support multi-level inheritance/polymorphism.

**Comment by jhclouse** [comment]
For what it's worth, here's my use case. I want to create a payment method for cart checkout. The mutation to create a payment method needs to take either a credit card or a bank account as input, but not both. So I'd like to have a single non-null field in the mutation input that represents both.

**Comment by joelgetaction** [comment]
I don't understand the point of having output union types but not input union types?  Like why would you have one without the other?  For example, suppose a user is uploading a photo with shapes drawn on top of it to annotate it?  we could have:
```graphql
type ShapeLine { ... }
type ShapeRectangle { ... }
type ShapeArrow { ... }
union Shape = ShapeLine | ShapeRectangle | ShapeArrow

type Photo {
  id: ID!
  url: String!
  shapes: [Shape!]!
}

input UploadPhotoInput {
  url: String!
  # The order of shapes matters because later shapes in the array will be drawn on top
  # of earlier shapes in the array
  shapes: [ShapeInput!]
}
type ShapeLineInput { ... }
type ShapeRectangleInput { ... }
type ShapeArrowInput { ... }
union ShapeInput = ShapeLineInput | ShapeRectangleInput | ShapeArrowInput
```

This is a perfect example of why we absolutely need union input types.  If we don't have union input types then I have to have separate arrays of lines, rectangles and arrows, but because the order of their drawing matters (so that we can capture what was drawn on to of each other) I now have to patch together some artificial and clumsy system of mixed ordinals across separate arrays of `ShapeLineInput,ShapeRectangleInput,ShapeArrowInput` to preserve draw order or I have to create a tagged union of inputs so that I can put them in the same array and use array ordinal as draw ordinal.

Moreover, validation of separate arrays and / or tagged unions using something like `Joi` becomes a LOT harder without input unions ...

This simple example is made really clumsy all because GraphQL doesn't have input unions.

Any suggestions?

**Comment by raderio** [comment]
Swagger also has it https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#anyof

**Comment by hlminh2000** [comment]
We have a json filter DSL that was designed with some polymorphism, where a certain field may be either an object, or an array. This feature would allow us to model this directly in the GraphQL schema. Currently, we are having to use the JSON scalar from `graphql-type-json` and validate manually, which means the required shape is not apparent to the API consumer through schema inspection. This proposal would allow us to model it directly in the schema and make this structure apparent to the consumer.

**Comment by acao** [comment]
It seems this a duplicate of #395 which is also seeing a lot of progress? And has some spec written? Maybe should bring the discussions together?

**Comment by fennecbutt** [comment]
I'd find this very useful too, I've been implementing an `event_name, event_payload` type mutation where event_payload is polymorphic based on the value of event_name.

At the moment I'm going to have to implement each as separate mutations, or event_payload will just have be "anything" in the schema, which will mean that I need to provide additional documentation for end users.

**Comment by raderio** [comment]
```
sealed class Shape
class Circle(val radius: Double) : Shape
class Rectangle(val width: Double) : Shape
```

There are 2 types of polymorphism:
1. Polymorphism by field value, aka discriminator
1.1 Discriminator embedded in object
```
[ { "type": "CIRCLE", "radius": 10.0 }, { "type": "RECTANGLE", "width": 20.0 } ]
```
1.2 Discriminator external
```
[ { "type": "CIRCLE", "data": { "radius": 10.0 } }, { "type": "RECTANGLE", "data": { "width": 20.0 } } ]
```
2. Polymorphism by filed name, aka union, or `anyOf` from Swagger
```
[ { "radius": 10.0 }, { "width": 20.0 } ]
```
Because only `Circle` class has `radius` field, first object from list will be deserialized in `Circle` class.

Also should be possible to do nested deserialization(multi level).

**Comment by paean99** [comment]
So @raderio, trying to convert your comment to Graphql practical terms. The use of:

- _typename (and so graphql unions) would be the 'discriminator' case (1.1?)
- custom scalar types would be the 'anyOf' case

**Comment by xmnlab** [comment]
hey everyone!  sorry if I am missing something related to this and previous discussions.

also I have started to use graphql some months ago .. so maybe my comments below would be very noob, so sorry about this.

I am here because we are trying to grapqhl with schema.org. Schema.org is very flexible, so for example we could have something like `author: String | [String!] | Person | [Person!] | Organization | [Organization!]`.

maybe a little bit different about what was discussed, but what if graphql implements something a `Metadata` type?  it could be a extensible type (user could extend that and add more fields) or this type could have a lot of defined fields that could cover a lot of use cases.

about the input union, as there is already `_typename`, maybe it could be used (instead of `type`) as discriminator or if could use a (new) Metadata type `__meta.type` could be used  .. or also something like `<template>` ( not sure how it could be add in the language without break its specification). 

also it could be implemented inside scalar types and we could define a string field like `String(blank:False)`. of course it seems we can use directives to extend this capability (as SHACL does) but maybe if the language could support that in its core would be very interesting.

again, sorry if this comment is not pertinent

**Comment by alamothe** [comment]
@xmnlab You should do something like:
```graphql
author: Author

union Author = AuthorString | AuthorList | Person | AuthorPersonList | ...

type AuthorString {
  value: String
}

type AuthorStringList {
  value: [String!]
}

type AuthorPersonList {
  value: [Person!]
}
...
```

While this might look cumbersome, it works really well in practice, and you don't need additional concepts introduced (that is, unless you need a union for an input type 🙂 )

**Comment by xmnlab** [comment]
Thanks for the feedback @alamothe ! 

yes  I will need union for an input type as well :)

**Comment by mike-marcacci** [comment]
Can we revisit exactly _why_ a `__inputname` or other type marker is required here at all? I am very much of the same opinion as @jayhasyee in #415, in that the server is defining the input types in the first place and has everything necessary to disambiguate.

After all, data is just data, and there is no extra information being conveyed by a special-case `__inputname` parameter that can't be conveyed by a parameter defined on the data itself.

**Can anyone provide a use case that _requires_ a `__inputname`?**

**Comment by benjie** [comment]
Here's a very simple example that wouldn't work without `__inputname` or equivalent

```graphql
input PersonInput {
  name: String!
  age: Int
}

input OrganizationInput {
  name: String!
  numberOfMembers: Int
}

inputUnion ContactInput = PersonInput | OrganizationInput

type Mutation {
  addContact(contact: ContactInput!): Contact
}
```

```graphql
mutation {
  addContact(contact: {name: "IBM"}) {
    __typename
    name
  }
}
```

If we require users to add a special property to the input types to disambiguate them (we called this a "discriminator" above), why should it not be a single standardised explicit `__inputname`?

**Comment by xmnlab** [comment]
hey everyone, not sure if it is relevant, but I could resolve my issue related to union of input type using https://github.com/Cardinal90/graphql-union-input-type with `__typename` as `discriminator`

**Comment by mike-marcacci** [comment]
@benjie I disagree with that – I did agree at first, but when I really think about _why_ GraphQL has a `__typename` in the first place, I am re-convinced that it doesn't belong.

My understanding of the GraphQL type system is that it is primarily _structural_ as opposed to _nominal_, and that type and interface names exist for very pragmatic reasons given the relationship between the server and the client:

The server is responsible for exhaustively defining the types used by the API; the client is primarily concerned with fetching _the data that it needs_ and should not need to be concerned with the presence of types or fields that it doesn't want. Let's take your example:

```graphql
type Person {
  name: String!
}

type Organization {
  name: String!
  members: [Person]
}

union Contact = Person | Organization

type Query {
  contacts: [Contact]
}
```

If I'm a client with a query like this:
```graphql
query {
  contacts {
    name
  }
}
```
...then there's no way for me to tell the difference between a `Person` and `Organization` without querying a potentially huge field, which _I may not want_. Because I'm the client and not the server, I don't have the ability to add a field to the schema for disambiguation! This is why `__typename` is so important: because it gives the client a way to resolve types without over-querying. However – don't be fooled into thinking that `__typename` comes without its own catches: a huge advantage of GraphQL's primarily _structural_ type system is that you can start by defining a concrete type `Contact`, which is generally safe turn it into an interface later on; when clients use `__typename`, however, this becomes a breaking change, since there is no longer a `Contact` but a `Person` and `Organization`.

---

Now let's look at the dynamics of input polymorphism from the server's perspective:

```graphql
input PersonInput {
  name: String!
}

input OrganizationInput {
  name: String!
  memberIds: [ID!]
}

input union ContactInput = PersonInput | OrganizationInput

type Mutation {
  addContact(contact: ContactInput!): Contact
}
```

In this case, the server is unable to distinguish between a `PersonInput` and an `OrganizationInput` if the latter is sent with null for `memberIds`. However, this schema should simply never be written! The server gets to specify its data requirements _by writing the schema_ just like the client specifies its data requirements _by writing the query_. In this case, we've simply written a bad schema: this is identical to a query failing to request a `__typename` in its query. It doesn't get one because it didn't ask for one! The solution here is simply to add a required enum `type` field onto each input – or any other kind of field that would be relevant.

I hope this makes sense; adding `__inputname` is not only unnecessary, but really ugly, and in my opinion really muddies the server/schema and client/query symmetry.

**Comment by treybrisbane** [comment]
@mike-marcacci 

>My understanding of the GraphQL type system is that it is primarily structural as opposed to nominal

Can you explain how you arrived at this conclusion? Because I'm not sure it's correct...

Let's assume you're right and that GraphQL's type system is indeed structural.
Under that assumption, say we have the following schema:
```graphql
schema {
  query: Query
}

type Query {
  fooBar: FooBar!
}

type FooBar {
  name: String!
}
```
Such a schema can be queried like so:
```graphql
query {
  fooBar {
    name
  }
}
```
If GraphQL's type system is structural, we should be able to change the definition of `FooBar` to be any other structurally equivalent type without having to change the query.
Suppose we change the schema to this:
```graphql
schema {
  query: Query
}

type Query {
  fooBar: FooBar!
}

union FooBar = Foo

type Foo {
  name: String!
}
```
Unfortunately, in this schema, `FooBar` is _not_ `Foo` despite being structurally equivalent; it is a distinct type unto itself. This is evident by the fact that you must use an inline fragment to access the `name` field of a FooBar in a query:
```graphql
# Error!
query {
  fooBar {
    name
  }
}

# We must do this instead
query {
  fooBar {
    ... on Foo {
      name
    }
  }
}
```
Thus, we have a contradiction, and so GraphQL's type system cannot be structural in nature.

---

I'm not sure I follow the rest of your argument, but one thing did stick out to me:
>...However, this schema should simply never be written!...
>...In this case, we've simply written a bad schema...

How do you define "bad" in your example case?

The reason I ask is because I've seen numerous valid use cases for explicitly discriminated input unions across three (including this one) separate input union threads. Are you suggesting that such use cases are somehow wrong?

**Comment by mike-marcacci** [comment]
@treybrisbane thanks for the thoughtful reply. While this is not essential to my argument against `__inputname`, I think I might not have adequately described the ways in which GraphQL is structural vs nominal. GraphQL is structural in that the type system is specifically designed to make the action of subdividing a type into multiple types a backwards-compatible change.

Here I'll elaborate on the example I used in the final notes above.

Say I have a concrete type in my schema:
```graphql
type Contact {
  name: String!
}
```

My app becomes really successful, but we don't have the ability to immediately update all clients, because many are native apps, which users update at different paces. This makes it important that all our API changes are backwards-compatible.

However, we realized that we need to separate our `Contact` into separate `Person` and `Organization` types. Fortunately, GraphQL can handle this! Our `Contact` can become an interface that is inherited by the new types:

```graphql
interface Contact {
  name: String!
}

type Person implements Contact {
  name: String!
}

type Organization implements Contact {
  name: String!
  members: [Person]
}
```
---

A more concise example:
```graphql
type Foo {
  hello: String!
}

type Query {
  foo: Foo!
}
```
can easily be replaced with:
```graphql
type Bar {
  hello: String!
}

type Query {
  foo: Bar!
}
```
without breaking any queries, because the _structure_ is our primary concern.

---

All that said, you are completely right that type names play an important role, and your example with a union is a perfect example. However, as I attempted to describe in my previous comment, I believe this is for the practical purpose of _allowing the client to disambiguate without requesting additional information_. This is because **the client states its data dependancies by writing the query**. On the other side, **the server states its data dependancies by writing the schema**.

A query or schema is "bad" when it doesn't request the data it needs.
- If a query needs to differentiate between two types within a union, it needs to request disambiguating information from the server. Because a query doesn't get to define the schema, this is done by specifying the type or interface name (effectively "shape" name) in a fragment, and if necessary requesting the `__typename` in the response. A query that fails to do this is a "bad" query, and it would naturally be difficult to use one.
- If a schema needs to differentiate between two **input** types within an input union, it needs to request disambiguating information from the client. Because a server does get to define the schema, this is done by defining types that are different in any meaningful way. A schema that fails to do this is a "bad" schema, and it would naturally be difficult to use one.

I think there is definitely a need to discriminate between input union types; I just think that this can be solved in the schema itself, as opposed to the spec.

**Comment by mike-marcacci** [comment]
As a counter to my own argument, I can imagine some scenarios where using an `__inputname` could be more ergonomic: it allows the same input type to be used both inside and outside a union without adding requiring disambiguating info when it isn't needed, but requiring it when it is needed.

```
input CatInput {
  name: String!
  lives: Int
}

input DogInput {
  name: String!
  fetches: Boolean
}

input union PetInput = CatInput | DogInput

input PersonInput {
  name: String!
  pets: [PetInput]
}

type Mutation {
  addPerson(person: PersonInput): Person
  addPetToPerson(personId: ID!, pet: PetInput): Person
}
```

Without the spec-defined `__inputname`, `Dog` and `Cat` would have to be written twice – one set with disambiguating characteristics and one without. This is a pretty compelling case for me, so I fully retract my objections!

**Comment by gustavopch** [comment]
I had a case where I thought input unions would be needed, but I found a solution that's neither awkward nor requires changes to GraphQL spec. So I thought I should share my experience:

It's an e-commerce. Each order has an array of items. Each item can either be a product item (represents a single product) or a combo item (represents a combination of products usually with a discount). They have different structures. When submitting an order, I thought I needed input unions so that I could have something like this:
```gql
inputUnion OrderItem = ProductItem | ComboItem

input OrderCreationParams {
  items: [OrderItem!]!
}
```
But I ended up doing this:
```gql
input OrderCreationParams {
  productItems: [ProductItem!]!
  comboItems: [ComboItem!]!
}
```
Then I just needed to merge `productItems` and `comboItems` in the resolver (using JS here):
```js
createOrder: async (parent, { params: { productItems, comboItems, ...rest } }) => {
  const params = { ...rest, items: [...productItems, ...comboItems] }
  ...
}
```
Simple and ellegant.

So I think it should really be analyzed whether the complexity that will be added to GraphQL by having input union types is really worth. Note that I just needed 1 extra line in the SDL and 1 extra line in the resolver to accomplish what I needed.

Also, it seems the case showed in the original post could easily be solved with this SDL:
```gql
type Mutation {
  createMedia(addMediaInput: AddMediaInput, addPostInput: AddPostInput): Media
}
```
And this resolver:
```js
createMedia: async (parent, { addMediaInput, addPostInput }) => {
  const input = addMediaInput || addPostInput
  if (!input) throw new Error('No input specified')
  ...
}
```
How awkward is that? Or did I miss something?

**Comment by binaryseed** [comment]
That works when you can accept a list of items, but it's not possible to use this when you can only accept a single item...

**Comment by gustavopch** [comment]
> That works when you can accept a list of items, but it's not possible to use this when you can only accept a single item...

Why? Look at my second example. It splits what would be a single argument into two and easily solves the problem.

**Comment by mplis** [comment]
When you split a list of heterogenous types into multiple lists of homogenous types, you lose the original ordering of items. If that ordering is not significant to your use case, then it's fine, but that's not going to be the case all the time.

**Comment by gustavopch** [comment]
> When you split a list of heterogenous types into multiple lists of homogenous types, you lose the original ordering of items. If that ordering is not significant to your use case, then it's fine, but that's not going to be the case all the time.

That's a good point. But would it be too much to require the client to pass an `index` property that will be used to reorder the merged array if the order matters?
```
createOrder: async (parent, { params: { productItems, comboItems, ...rest } }) => {
  const items = [...productItems, ...comboItems].sort((a, b) => a.index - b.index)
  const params = { ...rest, items }
  ...
}
```

**Comment by mplis** [comment]
> But would it be too much to require the client to pass some sort of an index property that would be used to reorder the merged array if the order matters?

Whether it's too much to ask likely depends on the use case, but I think it's definitely more awkward for the client than being able to send all the items in a single list.

**Comment by AlessandroFerrariFood** [comment]
> > That works when you can accept a list of items, but it's not possible to use this when you can only accept a single item...
> 
> Why? Look at my second example. It splits what would be a single argument into two and easily solves the problem.

In your example both are optional but you can't have a single item required.
Not none and not both.

**Comment by gustavopch** [comment]
@mplis OK, I agree there's some awkwardness on it. But if I understand the proposal correctly, it also adds an extra property `kind`. That's as awkward (if not more) than asking the `index` (or `position`, you name it). I say that because that `kind` is purely implementation detail while `index` can still be considered actual data. Also should clients assume that lists are always ordered lists? Isn't that an implementation detail as well? I mean, the list may be stored in the server as a hashmap keyed by ID or something. Who knows? I'm not sure the client should assume that the order will be preserved.

@AlessandroFerrariFood IMO it would be trivial to add a check like `if (x && y) throw new Error('Both x and y were passed. Expected only one of them.')`.

**Comment by frankdugan3** [comment]
@gustavopch The biggest issue with your solution from my perspective is that it defers something that could be statically validated to something validated at runtime server-side, and which must be implemented by the resolver which is not portable to other backend languages. I think increasing type safety within the spec itself is worthwhile.

**Comment by paean99** [comment]
That solution and many others are practical. But don't they miss the point of graphql as a typed query language, since these solution are implemented in the resolvers?

**Comment by gustavopch** [comment]
@frankdugan3 @paean99 I think you are right.

**Comment by mike-marcacci** [comment]
To chime in here again, I don’t think there is anything that is impossible with the current spec, since tagged unions are totally possible; I think the real goal here is better ergonomics and making invalid states unrepresentable. For example, if I want to accept a union type as an input argument, it’s quite trivial to use two named  arguments with different types. Similarly, a heterogeneous list can simply be a list of wrapper types where each concrete type is relegated to its own field (this is the tagged union approach talked about by @leebyron).

So the context of discussion here must be about making this feel less awkward, or making stronger type guarantees for tooling. The latter comes into play when we consider that the schema has no way of describing how only one of the fields can contain content at a time. There is a bit of a moving line in this argument, as it’s outside the goals of GraphQL to represent all validation logic in the schema (for example, username character restrictions are clearly outside the scope of concerns).

From an ergonomic standpoint I am not totally convinced that the trade-offs in this proposal make for a cleaner API in general; they just make it different and I have yet to wrap my head around the consequences.

When I started writing Golang, I was really frustrated by the lack of genetics and overloading, and it felt dirty to write large numbers of very similar functions and structs. My experience with other languages set me up to fight go’s type system. But once I wrapped my head around and embraced the go idioms, they no longer seemed dirty or hacky, and the language worked quite well. I do feel that maybe some of the advantages in this proposal are simply changes to the mental model, which is not necessarily broken in the first place.

I don’t have super strong opinions either way, but wanted to express my concerns nonetheless.

**Comment by paean99** [comment]
My personal opinion is similar to @mike-marcacci.
That is why i think this thread is so good, since it shows so many valid arguments for both sides.

> There is a bit of a moving line in this argument, as it’s outside the goals of GraphQL to represent all validation logic in the schema (for example, username character restrictions are clearly outside the scope of concerns).

As an aside i would just like to point that graphql already has that functionality with Scalar types, if one so desire. It is just a question of defining our types a little more precisely.

**Comment by gustavopch** [comment]
@frankdugan3 @paean99 I'm watching some talks in YouTube about GraphQL schema design. I think I'm seeing a consensus that polymorphism is not a good thing.

So instead of a single `createMedia` mutation that would accept different forms of inputs as the original post wants, the solution would be to have specific mutations for each kind of input: `createPost` and `createImage`.

Even if the client UI would present both things as a single form to the user, I think it makes sense to treat them as different things on the realm of GraphQL.

**Comment by bgentry** [comment]
@gustavopch while there are certain simple scenarios for which separate GraphQL endpoints make sense, there are others where that would result in a crazy amount of API bloat. How would you address the IFTTT use case [I commented on previously](https://github.com/graphql/graphql-js/issues/207#issuecomment-276852250)? Multiple separate mutations for each of the hundreds of supported services?

**Comment by gustavopch** [comment]
> @gustavopch while there are certain simple scenarios for which separate GraphQL endpoints make sense, there are others where that would result in a crazy amount of API bloat. How would you address the IFTTT use case [I commented on previously](https://github.com/graphql/graphql-js/issues/207#issuecomment-276852250)? Multiple separate mutations for each of the hundreds of supported services?

Could you post an example of how your schema would be if GraphQL had input union types?

**Comment by frankdugan3** [comment]
@gustavopch There is already polymorphism in GraphQL, this issue here is the incongruity between types and input types. If the consensus is that polymorphism is bad, then we should get rid of `Interface` and `Union`. I don't see that happening... So since those constructs are here to stay, I think it makes a lot of sense to simplify mutations by adding similar polymorphism for input types.

**Comment by acao** [comment]
If folks want to see a demo of `inputUnion` similar to what I showed in the WG meeting last week, here ya go:

https://github.com/acao/apollo-graphql-input-union-example

I'll set up a live demo soon, with a working schema explorer

**Comment by alamothe** [comment]
Amazing work! :-) my vote goes for `__inputname` proposal. It's the cleanest approach (as in, transparent to the user) and consistent with how output types work.

**Comment by akomm** [comment]
@mike-marcacci https://github.com/graphql/graphql-spec/issues/488#issuecomment-487208421

To add to @treybrisbane's example, here on the mutation & input side:

```gql
input TestAInput {
  valueA: String!
  valueB: String!
}

input TestBInput {
  valueA: String!
  valueB: String!
}

type Mutation {
  testA(input: TestAInput!): TestAPayload!
  testB(input: TestBInput!): TestBPayload!
}
```
Try run this (TestBInput passed to testA which requires TestAInput):

```gql
# Variable type 'TestBInput!' doesn't match expected type 'TestAInput!'
mutation IsNotStructural($input: TestBInput!) {
    testA(input: $input) {
        ... on TestASuccessPayload {
            # ...
        }
        ... on TestAFailurePayload {
            # ...
        }
    }
}
```
Additionally, consider this is also an error:

```gql
# Unknown field "valueC" on input type "TestAInput". Did you mean "valueA", or "valueB"?
mutation IsNotStructural {
    testA(input: {valueA: "a", valueB: "b", valueC: "c"}) {
        ... on TestASuccessPayload {
            # ...
        }
        ... on TestAFailurePayload {
            # ...
        }
    }
}
```

For me graphql looks very nominal.

Your [example](https://github.com/graphql/graphql-spec/issues/488#issuecomment-487254283) has nothing to do with structural typing and is just a lucky case in which your new types have a common shape and thereforce you can swap it with a common interface.

I am strongly for input unions, if needed with `__inputname`. I have cases where input variants are mutually exclusive. I have to create different Payload/Input types and mutation fields to represent all those possible variants. Even if you write some abstraction over the type system which generates all those types for you, the whole system is polluted with tons of similar named types.

**Comment by benjie** [comment]
@akomm I'd be interested to know if https://github.com/graphql/graphql-spec/pull/586 solves your requirements? (Please discuss in that thread.)

**Comment by akomm** [comment]
@benjie I was initially rather with the `inputUnion` approach following this discussion. But after reading your post and thinking about it I found issues with `inputUnion`. Additionally, while you focus on the input topic there, in my opinion `@oneField` can even be interesting in context of output types. I am currently using unions for mutation success and failure outcomes (failures representing user errors, not exception, etc.). I have to utilize `__typename` for this and this does not feel as a good solution to me. With `@oneField` I could return to an older approach I had, where I could simply have `success` and `failure` fields which are never non-null at once. I did not like it because the mutual exclusiveness was not guaranteed by the type system.

**Comment by benjie** [comment]
@akomm Indeed, enabling it to cater to certain output situations was something I kept in mind whilst writing the PR. I don't want to start that discussion yet though, I feel it would muddy the waters (and the input use case is a lot more compelling IMO). I'm glad you spotted that too though!

**Comment by akomm** [comment]
@benjie

What we actually have then with your proposal is similar to `algebraic enumeration`, like union, however union semantics have the issue that you loose the type information at runtime when you receive data, making the need of an `__typename` property to transport it.

WIth the tagged union the variants of the enumeration are transported with the data and accessible by client, therefore no `__typename` or `__inputname ` is needed.

However currently no new keyword/type is introduced, simply because enum already exists non algebraic and changing this would probably be a BC and involve more work and more time to introduce a well working solution, than the annotation. And I think this semantic thing is something that throws some people off. 

I mean the `@oneField` might also be potentially a `@enum`, however I see there might be confusion with the `enum` that already exists.

**Comment by shadrech** [comment]
A lot of discussion on this subject. Any code merged?

**Comment by alamothe** [comment]
One obvious way to do this is pack everything inside a JSON, and pass it as a string parameter. The JSON should have `__inputname` where appropriate. Then on the server just parse and process it as if it was a union. If using TypeScript, it's easy to define proper types and get type checking. The only drawback is that the type is not visible in the GraphQL schema. Works for me 🙂

**Comment by nuttycom** [comment]
There have been a couple of comments along the way suggesting that tagged union types are not supported in typed languages like Java. This is incorrect - tagged unions (sum types) are represented in these languages using the Visitor pattern (the corrected version thereof: http://logji.blogspot.com/2012/02/correcting-visitor-pattern.html)

**Comment by Vadorequest** [comment]
This RFC was opened 14 months ago, and this limitation has been limiting GraphQL users for years now. Something should have been done by now, type polymorphism is a basic feature in many modelling tools and it's limiting all tools built using GraphQL.

At the very least, could we have a "status summary" in the main post, so that it's easier to follow this thread? As some other people pointed out, there has been extensive discussion in the thread and it's hard to follow.

In this case (that seems complex), let's not forget that `"best" is often worse than "good enough"`. 

Something needs to happen. This has been/is being implemented by GraphQL developers over and over. 
Prisma is currently working on their own RFC for instance because the official RFC feels stuck.

**Comment by dendrochronology** [comment]
It'll actually be 4 years next week, if you count graphql/graphql-js#207 😓 

I agree with @Vadorequest . I appreciate the thoroughness of the discussion here, but I fear that perfect has become the enemy of done.

This seems to be a serious weakness when it comes to modeling real-world, non-trivially complex data structures, and making reasonable GraphQL APIs for accepting said structures. As such, we resort to workarounds like: giant JSON blobs, ugly/hack-ish/insanely verbose GQL schemas, patching/customizing core GraphQL tools, or going around GraphQL entirely and side-loading data via REST, etc. My team has done all of those things at some point, and none of them are much fun. 

I realize that there are tradeoffs with all the proposed solutions, but what we're talking about is somewhat advanced usage, so I don't mind doing a bit more work on the backend when it comes to coding resolvers and mutations, like handling a discriminant field, or needing to understand a bit more about possible shapes of data. That may not be perfectly clean, but neither is the real world 😛

I'd love to see the spec team make an executive call on how to handle this, so folks in the community can plan accordingly.

**Comment by benjie** [comment]
This topic is actively discussed by the working group; here's notes from some recent WGs:

- https://github.com/graphql/graphql-wg/blob/b51415033fc7c6f5dc0d4184603d75bf1fcd0d17/notes/2019-10-10.md#review-input-union-rfc-and-discuss-the-next-steps-30m-vince
- https://github.com/graphql/graphql-wg/blob/master/notes/2019-09-12.md#input-union-vince-15m
- https://github.com/graphql/graphql-wg/blob/master/notes/2019-07-03.md#previous-action-items
- https://github.com/graphql/graphql-wg/blob/master/notes/2019-06-06.md#onefield-directive-input-unions--benjie-gillam-benjie
- https://github.com/graphql/graphql-wg/blob/master/notes/2019-05-02.md

We agree that `"best" is often worse than "good enough"` (aka `the perfect is the enemy of the good`). We are actively working on the Input Union RFC document to help us analyse the various goals and trade-offs that are at work; the document exists here: https://github.com/graphql/graphql-spec/blob/master/rfcs/InputUnion.md and has a few open PRs to it right now, including:

- https://github.com/graphql/graphql-spec/pull/626
- https://github.com/graphql/graphql-spec/pull/625
- https://github.com/graphql/graphql-spec/pull/623

We all feel this pain, which is why we're actively working to solve this particular issue. There is no perfect solution (that we have found) so we are all working towards being able to pick a solution that is acceptable to everyone, even if it's not perfect for all use cases.

You can get involved by contributing to the input union RFC document with your use-cases, aims, and the trade-offs that you would like to avoid. You can also become involved with the working group to help move the conversation on.

@binaryseed, who is leading this charge, has proposed in #627 that this and many other of the input union/input interface/input polymorphism issues and pull requests are closed in favour of more focussed discussion around the RFC document itself.

You may also be interested in reading [the "Guiding Principles" of the GraphQL spec](https://github.com/graphql/graphql-spec/blob/master/CONTRIBUTING.md#guiding-principles) as that will help to understand how we are comparing the various options.

**Comment by IvanGoncharov** [comment]
We have begun building a single RFC document for this feature, please see:
https://github.com/graphql/graphql-spec/blob/master/rfcs/InputUnion.md

The goal of this document is to consolidate all related ideas to help move the proposal forward, so we decided to close all overlapping issues including this one.
Please see discussion in #627

If you want to add something to the discussion please feel free to submit PR against the RFC document.

**Comment by venkatg2** [comment]
GraphQLError: Syntax Error: Unexpected Name "inputUnion"
i am getting this error..

**Comment by gustavopch** [comment]
@venkatg2 It's still an RFC, not implemented yet.

**Comment by venkatg2** [comment]
@gustavopch  ah thanks. any idea when it will be implemented?.. is there any alternative?

**Comment by gustavopch** [comment]
@venkatg2 I don't know when it will be implemented. I suggest you read all the previous comments on this issue to look for some ideas on alternative approaches.

**Comment by maka-io** [comment]
Sadly, I landed here today to find yet another "against the grain" GraphQL short-coming. Hopefully some day (some year?) GraphQL Mutations will be intuitive as their Query counterpart.

Back to overloading Query to do what Mutations can't do 😞

**Comment by sfalihi** [comment]
i want to use union-INPUT for this purpose:

```
input input_user {
     firstname: String
     lastname: String
}
 selectUser(input_user | AND: [input_user] | OR: [input_user]) 

```
** this way i can do all sql and/or operator with graphql nicely!

**Comment by acao** [comment]
@donalatorre the input union spec was updated only 7 days ago!
https://github.com/graphql/graphql-spec/blob/master/rfcs/InputUnion.md

there is a new, special input union working group that is actively working on this we speak. a consensus was finally reached after years of deliberation. its such a complex and contentious issue, that a seperate working group was needed, because it was taking over the working group calls!

i wouldn't be surprised if it ships with graphql-js@16. If you're interested and able, I'd highly reccomend getting involved to help advance it, anything from adding more use cases, to helping with reference implementations, etc.

**Comment by michaelstaib** [comment]
@donalatorre there has been a lot of work done. @benjie has put a tone of time into this. The approach at the moment is called the tagged type.

**Comment by acao** [comment]
Tagged type proposal is even more flexible and requires less tooling changes :) see

https://github.com/graphql/graphql-spec/issues/488#issuecomment-705188743

**Comment by gmac** [comment]
For those seeking alternatives, please consider `@oneOf` – https://github.com/graphql/graphql-spec/pull/825. I have yet to encounter a situation that couldn't be solved as well or better using that construct, and it works today with a simple server-side validation.

**Comment by bogdibota** [comment]
If, like me, you need a quicker fix than `@oneOf` – https://github.com/graphql/graphql-spec/pull/825, here you go:
```typescript
export const UploadOrString = new GraphQLScalarType({
  name: 'UploadOrString',
  description: 'Upload | String',
  serialize() {
    throw new Error('UploadOrString is an input type.');
  },
  parseValue(value) {
    if (typeof value === 'string') return value;
    else return GraphQLUpload.parseValue(value);
  },
  parseLiteral() {
    throw new Error('UploadOrString.parseLiteral not implemented');
  },
});
```
an in `.gql`:
```gql
scalar UploadOrString

input ProductInput {
  # for new images: the File object
  # for exiting images: the Id of the image
  # for removing images: skip from the list
  images: [UploadOrString!]!

  # ...
}
```

**Comment by brilovichgal** [comment]
Hi, 
I cant use inputUnion, it is writing to me that "Syntax Error: Unexpected Name "inputUnion"" what should i do?

**Comment by gmac** [comment]
@brilovichgal InputUnion was only a proposed spec, it’s not actually valid GraphQL. Read up this thread for other ideas.

**Comment by brilovichgal** [comment]
> @brilovichgal InputUnion was only a proposed spec, it’s not actually valid GraphQL. Read up this thread for other ideas.

So there is no real solution for it?
All the suggestions don’t work..

**Comment by gmac** [comment]
@brilovichgal dunno… IMO, `@oneOf` works considerably better than the idea of input union ever could. It circumvents the type identity issues inherent with the input union concept (multiple possible inputs with similar fields… what type is the input?), and it allows for mixed input types (submit objects, scalars, or lists); it’s also more expressive, and uses plain GraphQL that works today. Just add a simple server-side validation and you’re up and running. Best of luck.

**Comment by Vinccool96** [comment]
I extremely object with the decision to use `@oneOf`. My use-case has me wanting to accept between `{ recurrenceType: RecurrenceType! }`. `{ recurrenceType: RecurrenceType!, endDate: Date! }`, and `{ recurrenceType: RecurrenceType!, ocurrences: Int! }`. Using `@oneOf` just isn't possible.

**Comment by benjie** [comment]
```graphql
input RecurrenceOption @oneOf {
  typeOnly: RecurrenceType
  typeAndEndDate: RecurrenceTypeAndEndDate
  typeAndOccurrences: RecurrenceTypeAndOccurrences
}
input RecurrenceTypeAndEndDate {
  recurrenceType: RecurrenceType!
  endDate: Date!
}
input RecurrenceTypeAndOccurrences {
  recurrenceType: RecurrenceType!
  ocurrences: Int!
}
```

**Comment by gmac** [comment]
@Vinccool96 your specific circumstance is only suited to input union potential because all of your input types happen to have a unique composition of required fields. With optional fields in play, the backend cannot reliably disambiguate input type identities, which is where input union falls apart. The oneOf pattern is a pretty robust solution, and also solves the problem of allowing mixed input across scalars, lists, and input objects.

**Comment by benjie** [comment]
Since all types have the `recurrenceType: RecurrenceType!`, an alternative solution for @Vinccool96's use case could be:

```graphql
input RecurrenceOption {
  recurrenceType: RecurrenceType!
  extra: RecurrenceExtra
}

input RecurrenceExtra @oneOf {
  endDate: Date
  occurrences: Int
}
```

> Using `@oneOf` just isn't possible.

Since this statement doesn't seem to hold, please could you clarify what your issue with the solutions above are, or what you think the solution should be?
